### PR TITLE
Initial R API

### PR DESCRIPTION
**Pull request by aaronwolen**
This copies over some of the initial work on the tiledbsoma R API. Other than providing the package skeleton, it also includes:

- refactored R6 base classes for interacting with TileDB groups and arrays
- `SOMASparseNdArray` class and unit tests
- various internal helper utilities

### CODE REVIEW

**Review by johnkerl — APPROVED** [review]
🚢

File: apis/r/R/SOMASparseNdArray.R [hunk 1]
```diff
@@ -0,0 +1,118 @@
+#' SOMASparseNdArray
+#'
+#' @description
+#' `SOMASparseNdArray` is a monomorphic, sparse, N-dimensional array with offset
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply apis/r/R/SOMASparseNdArray.R:4]: Good call. This wording was taken from the spec at the time. Subsequently we discussed that monomorphic was ambigous and decided to remove, so I'll do the same here.

File: apis/r/R/SOMASparseNdArray.R [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+#' (zero-based) integer indexing on each dimension. The `SOMASparseNdArray` has
+#' a user-defined schema, which includes:
+#'
+#' - type - a `primitive` type, expressed as an Arrow type (e.g., `int64`, `float32`, etc)
+#' - shape - the shape of the array, i.e., number and length of each dimension
+#'
+#' All dimensions must have a positive, non-zero length.
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply apis/r/R/SOMASparseNdArray.R:11]: This language was also taken from the spec and is still current. We should revise it there first if we want to make a change.

File: apis/r/R/SOMASparseNdArray.R [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+SOMASparseNdArray <- R6::R6Class(
+  classname = "SOMASparseNdArray",
+  inherit = TileDBArray,
+
+  public = list(
+
+    #' @description Create a SOMASparseNdArray named with the URI.
+    #' @param type an [Arrow type][arrow::data-type] defining the type of each element in the array.
+    #' @param shape a vector of integers defining the shape of the array.
+    create = function(type, shape) {
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply apis/r/R/SOMASparseNdArray.R:32]: Right now `type` is consistent with the [python api](https://github.com/single-cell-data/TileDB-SOMA/blob/de1fa8cec569bc03f6a6bc7f04cdf5941bc8138f/apis/python/src/tiledbsoma/soma_sparse_nd_array.py#L38-L41). I'm up for renaming it later in both places.

File: apis/r/R/SOMASparseNdArray.R [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        filter_list = zstd_filter_list
+      )
+
+      # array schema
+      tdb_schema <- tiledb::tiledb_array_schema(
+        domain = tiledb::tiledb_domain(tdb_dims),
+        attrs = tdb_attr,
+        sparse = TRUE,
+        cell_order = "ROW_MAJOR",
+        tile_order = "ROW_MAJOR",
```
[the remark this answers is not in the document]
Comment (aaronwolen) [reply apis/r/R/SOMASparseNdArray.R:70]: Indeed!

**Pull request merged**